### DISCUSSION TITLE
Simple Joomla CCK

### DISCUSSION THREAD

**Discussion by Llewellynvdm** [joomla/joomla-cms#39461] [RFC]
## This RFC has passed the official entrance vote in the production department.

- This feature is eligible for the Core.
- A working group will be formed to work out the implementation details through this RFC process.
- Official RFC Documentation can be found here: https://github.com/joomla/rfc/pull/26

# What Next

- Reading over the comments made here:  https://github.com/joomla/rfc/pull/26 please continue the discussion here on this thread.
- [Give more Use Cases here](https://github.com/joomla/joomla-cms/discussions/39461#discussioncomment-4468584).

### 1. Summary

A “Simple CCK” system for extending the data model of whatever type you are using and providing extended query and display functionality (such as adding more query filters). Just a CMS extended with filter and ordering (by menu link or module with switch, etc) and some nice view overrides.

For example, if you have a list of activities to do on weekends. With fields such as "date", "free or paid", "city". You should be able to filter by activities in a specific city. And to be able to sort them by price or by date.
In resume, improve the Joomla Content as a Simple Content Construction Kit (but not a real or complex “Advanced CCK”, with real content-type management, logical relation between them and for creating web apps (like Flexi, Seblod, Fabrik)

**2. Why Bother?**

Actually, if you just want a CMS extended with filter and ordering, you need to use a third party CCK and stop using the Joomla Core Articles, which prevents using the vast majority of Joomla's own utilities (like Joomla Core Tags, Joomla Core Custom Fields, Joomla Core Categories, Joomla Multilanguage and Associations, Joomla Workflows, etc.) as well as the vast majority of other third-party extensions.

Adding some order and filter option based on the custom fields will help creating Content website, keeping the Joomla! spirit without bringing the CCK difficulties!

See https://github.com/joomla/joomla-cms/issues/26258

**Comment by Llewellynvdm** [comment]
### What’s your use case?

If this feature sounds helpful to you, please reply to this comment with your use case. The more details and examples, the better we can understand your needs. Thank you!

**Reply by SemaphoreOxalis** [reply]
**About ordering in front for core content type (articles, contact etc...)**
A classic one : 
- creating a category for events or news
- add a least one custom fields date because published and modify are not usefull for definingan event content type 
- Try to create a menu link for a blog views ordered by this field from the core menu link : you can't.
- Same issue/limits for the article modules mod_articles_xxxx
We don't try to create a full agenda component here, just tryning to order soomple simple content by a customfield. SOmetimes we don't need a complete extention.

We need to be able to have ordering(s) option(s) based on all core (already done) and custom fields.

**Filtering are a complete other function** because it's way bigger 
It could be simple to write this like "as a weblasster I want to be able to add filters based on each core and custom field for the core list view (list, blog etc) in order to pre-filter the content on the frontview" 
buuuut
a complete filter function will permit to 
- use multiple filters 
- with boolean conditions (or, and, not...) for setting a filter like customfields1 value is NOT zzzzz
- and test them against some global var like now(), url params etc
....
So maybe just a free input with a filtering syntax "à la " articleanywhere for power users ^^

Because I quote it : ArticleS Anywhere is THE filtering/ordering best example I can imagine and we use it but it's friustrating to re-code a blog view with AA for just having an ordering option.

**Comment by brianteeman** [comment]
> In Joomla 3.9 ([PR#20890](https://github.com/joomla/joomla-cms/pull/20890)), custom admin menus with a lot of **
Filtering** options were added, which is another major step towards a simple CCK supporting (kind of) different content
types.

As I wrote that PR (and didnt apply it to other components as I had promised) let me point out the outstanding issues with that which can hopefully be improved upon for the admin.

Basically that PR is a hack that lets you create a link to the article list with a series of pre-applied filters and because of those filters when you create a new article those filters are also applied

![image](https://user-images.githubusercontent.com/1296369/209000331-0306632e-742a-4ce6-89a6-20c70c1331c5.png)

![image](https://user-images.githubusercontent.com/1296369/209000474-d53af301-0e01-43c8-8532-7cd1874f0586.png)

![image](https://user-images.githubusercontent.com/1296369/209000594-b793927f-8e8c-4977-bb9f-85dc51ffc1bd.png)

The three obvious (to me at leats) areas that need to be imporved are

1. The title is always "Articles" - it should be the menu item name etc - otherwise there is no difference between multiple menu links with different filters
2. The filter bar is always open in joomla if there are filters selected. When used in this way then the filter bar should probably be closed and only open if additional filters are applied. Also a preselected filter can be changed in the list. I think it is better and more cck like if those pre-selected filters can not be changed
3. As with point 2 above the pre-selected filter (category in the screenshot) should not be able to be changed.

**Comment by TLWebdesign** [comment]
isn't what we want just a native filter option on category blog views? (at least that would be my main use case) A way to set up filters for the articles based on either custom fields or other article related information. At the moment if i need a solution like this i would implement i use isotope filtering in a custom layout override https://isotope.metafizzy.co/filtering.html. Or i've done a bit of js to filter based on data tags in the article div.

**Reply by micker** [reply]
Yes filter by custom field need to be added
Actual custom field are only for display but need to be consider liké a value....
Search by text in field
Filter by date
Filter by value
Filter between x values
With some filter option like a dropdown search input, slider, radio, checkbox
That can finish and extend custom field feature

**Comment by peteruoi** [comment]
I ping some from the discusion [https://github.com/joomla/joomla-cms/discussions/19150]([https://github.com/joomla/joomla-cms/discussions/19150](https://github.com/joomla/joomla-cms/discussions/url))  that they were in favour of the cck cause it has been accepted and may be interested to give here their opinions.
@xmanflash @ssnobben  @artur-stepien @helpwise @pulsarinformatique @leoalv

**Reply by helpwise** [reply]
Thanks Peter. The original discussion was here from memory.  https://github.com/joomla/joomla-cms/discussions/19150  (that other link doesn't click through)

I had a very clear idea of how this should work and be integrated with the Joomla system and to me it seemed so simple, as to not do it would be madness.  You can see how the conversation went in that GitHub thread "Code it yourself and submit it, we are sick of people asking for stuff" was basically the response.  

The CCK - done well - could revitalise Joomla.  Joomla really needs a selling point at this late hour.  It might be worth a shot.

The CCK Types feature should link ACL, Custom Fields, Menus, Ajax Filters, have multiple categories (under them) and all be easily managed from the Admin via dropdowns for Customer Use. 

Its almost a year since the last post - what has been done so far ?

Here is the image I mocked up for the user interface in Admin.  Its in that linked thread along with a text description.

![joomal_types](https://github.com/joomla/joomla-cms/assets/29443705/6167adb1-5a8c-4b12-a78a-f83ff089480b)
